### PR TITLE
Using Key Store Manager as a OSGI service

### PR DESCRIPTION
**Pull request by ruwanta**

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/internal/SAMLSSOAuthenticatorServiceComponent.java [hunk 1]
```diff
@@ -38,12 +39,15 @@
...
  * @scr.component name="identity.application.authenticator.samlsso.component" immediate="true"
+ * @scr.reference name="private.key.provider" interface="org.wso2.carbon.identity.core.KeyProviderService"
+ * cardinality="1..1" policy="dynamic" bind="setKeyProvider"  unbind="unsetKeyProvider"
  */
 
 public class SAMLSSOAuthenticatorServiceComponent {
 
     private static Log log = LogFactory.getLog(SAMLSSOAuthenticatorServiceComponent.class);
     private static RealmService realmService;
+    private static KeyProviderService keyProviderService;
```
Comment (GayanM) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/internal/SAMLSSOAuthenticatorServiceComponent.java:50]: why keyProviderService is static ?

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/internal/ServiceReferenceHolder.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+package org.wso2.carbon.identity.application.authenticator.samlsso.internal;
+
+import org.wso2.carbon.identity.core.KeyProviderService;
+
+/**
+ * Holds OSGi service references for utility classes.
+ */
+public class ServiceReferenceHolder {
```
Comment (GayanM) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/internal/ServiceReferenceHolder.java:26]: Is this usable any where ?

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java [hunk 3]
```diff
@@ -28,26 +30,102 @@
...
+    public X509CredentialImpl(String tenantDomain, String idpCert, KeyProviderService keyProviderService)
+            throws SAMLSSOException {
+        this.keyProviderService = keyProviderService;
+        X509Certificate cert = null;
+
+        /**
+         * If IDP cert is passed as a parameter set the cert to the IDP cert.
+         * IDP cert should be passed when used with response validation.
+         */
+        if (idpCert != null && !idpCert.isEmpty()) {
```
Comment (GayanM) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java:79]: can use StringUtils

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java [hunk 4]
```diff
@@ -28,26 +30,102 @@
...
+            }
+            if (cert == null) {
+                throw new SAMLSSOException(
+                        "Cannot find the certificate from IdP certificate: " + idpCert + " , for tenant: "
+                                + tenantDomain);
+            }
+        } else {
+            PrivateKey key = null;
+            try {
+                if (keyProviderService != null) {
```
Comment (GayanM) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java:93]: need to handle the situation of keyProviderService is null

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java [hunk 5]
```diff
@@ -28,26 +30,102 @@
...
+                    if (log.isDebugEnabled()) {
+                        log.debug("Using Key provider service to lookup the private key and certificates");
```
Comment (GayanM) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java:95]: log the tenant domain

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java [hunk 6]
```diff
@@ -28,26 +30,102 @@
...
+                    if (certificateFromService instanceof X509Certificate) {
+                        cert = (X509Certificate) certificateFromService;
+                    }
+                }
+
+                if (key == null) {
+                    throw new SAMLSSOException(
+                            "Error retrieving private key from keyProviderService for tenant " + tenantDomain);
+                }
+                if (key == null) {
```
Comment (GayanM) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java:108]: Why duplicate null checks, second one should be "cert"

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java [hunk 7]
```diff
@@ -28,26 +30,102 @@
...
+                    throw new SAMLSSOException(
+                            "Error retrieving the X.509 Certificate from keyProviderService for tenant "
+                                    + tenantDomain);
+                }
+                if (log.isDebugEnabled()) {
+                    log.debug("Key provider service was able to find the private key:" + key + " and certificate:"
```
Comment (GayanM) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java:114]: log the tenant

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java [hunk 8]
```diff
@@ -28,26 +30,102 @@
...
+     * Instantiates X509Credential.
+     * If the IDP cert passed is not null the instantiated credential object will hold the passed
+     * cert and the public key of the cert.
+     * Otherwise the object will hold the private key, public key and the cert for the respective
+     * tenant domain.
+     *
+     * @param tenantDomain tenant domain
+     * @param idpCert      certificate of the IDP
+     * @param keyProviderService the key provider service
+     * @throws SAMLSSOException
```
Comment (kasunbg) [inline components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java:68]: missing description.

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/internal/SAMLSSOAuthenticatorServiceComponent.java (hunk 1 above)
Comment (GayanM) [reply components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/internal/SAMLSSOAuthenticatorServiceComponent.java:50]: no need here as service reference already has it

File: components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java (hunk 4 above)
Comment (GayanM) [reply components/org.wso2.carbon.identity.application.authenticator.samlsso/src/main/java/org/wso2/carbon/identity/application/authenticator/samlsso/manager/X509CredentialImpl.java:93]: not a problem exception will be thrown

**Comment by GayanM** [comment]
comments fixed with https://github.com/wso2-extensions/identity-outbound-auth-samlsso/pull/45

**Pull request merged**